Provide source location for declared objects

SkySkimmer opened the pull request:
cf https://github.com/coq/vscoq/pull/911

Overlays:
- https://github.com/LPCIC/coq-elpi/pull/718

ejgallego commented on vernac/declare.ml:
```diff
@@ -566,7 +567,22 @@ let is_unsafe_typing_flags flags =
+    ~cache:(fun (kn,loc) ->
+        loc_table := Cmap_env.add kn loc !loc_table)
+    ~subst:(Some (fun (subst,(kn,loc)) -> Mod_subst.subst_constant subst kn, loc))
+    ~discharge:(fun x -> Some x)
```
Note the discussion on https://github.com/coq/coq/pull/16261/ , in particular this approach requires fully loading a `.vo` to determine location for objects, so it seems to rule out many interesting use cases.

Note also that as of today, .glob files work well for jump to definition, coq-lsp 0.2.0 implements global jump to definition just fine using this approach.

A flaw both approaches have is that whitespace-only changes will alter the hash of the file, someone complained about this, but IMO fairly orthogonal.

[Also you can likely reuse the `loc` parts of #16261]

SkySkimmer replied:
Using glob files does sound like a good alternative.
Do you use them to find the references or just the definition points? 
Have you had issues with missing entries?
What do you do for definitions from the current file?

SkySkimmer replied:
cc @rtetley

ejgallego replied:
@SkySkimmer it works pretty well so far, didn't do a super rigorous test. For definitions in the current file, they are processed as they are parsed, anyways you want to do that for more meta-data (display doc, etc...)

I mean, you can have this new ad-hoc table that is kept using the `unit -> unit` libobject side effect, but it is neither useful nor very maintenable IMHO.

gares replied:
We did use .glob files in Coqoon for the same reason, but that is a bit ugly because:
- glob files may not exist (eg -no-glob)
- glob files may not be installed
- the current file does not have a glob file (yet). Worse:
  - you can't write a glob file since in interactive mode you undo/re-execute, but files on the disk don't roll back
  -  so we were using the possibility to stream the glob info for the current file to the feedback channel and we were storing this info UI side, keep it updated on changes, etc

The approach here does not have the problems above, i.e. it works the same for other files or the current file and is kept in sync via the summary.

At the same time if you have executed a sentence, you also have already loaded all the .vo files that sentence needs, especially in continuous mode. So the advantage of not requiring to load .vo files seems marginal to me.

SkySkimmer commented:
Moved the table to nametab (seems appropriate to avoid having 1 table for each of constant / inductive / abbreviation, in the future we will probably want a table for every nametab entry)
added locations for abbreviations inductives and constructors

SkySkimmer commented on vernac/vernacinterp.ml:
```diff
@@ -106,9 +106,14 @@ and vernac_load ~verbosely entries =
-    ~unfreeze_transient:Vernacstate.Synterp.unfreeze
-    (fun () -> interp_expr ?loc ~st cmd)
+  Util.try_finally (fun () ->
+      Loc.set_current_command_loc loc;
```
We keep the current command loc in a global ref (put in loc.ml because I had no better idea) to be a backup when a command does not pass more precise loc info to declare.

ejgallego replied:
Can we please instead fix the commands to pass info as done in #16261 may provide some more places to fix (I didn't check tho).

The way I see it is that the hack is a recipe for problems, on the other hand, if a command didn't send its location, it can be fixed easily.

SkySkimmer replied:
Not necessarily easy, typically `abstract` has no good location at hand

also I don't see the problem so don't want to spend much time removing it

ejgallego replied:
Abstract could use the location of the tactic, couldn't it?

Actually if the hack was more local it would be OK, but the way it is is IMO problematic. If we put `loc` in a global variable, why for example not to put poly, and other arguments to cinfo in global variables too?

Design-wise, these are the main problems I see:
- the global variable may look fine today, but in the end, it will create maintenance problems if for example `vernacinterp` is refactored in a way more friendly for Flèche
- `vernacinterp` knows nothing about the command at hand, so what the PR is doing now is an _approximation_ of the location; this is limiting for LSP for example (see the current jump to def protocol spec)
- moreover I don't think fixing this hack should take a lot of time, as the previous PR attempting this shows

SkySkimmer replied:
>Abstract could use the location of the tactic, couldn't it?

no, the tactic may be in some `Ltac` definition

ejgallego replied:
Indeed to fix this properly (IMO) would require a lot of refactoring that indeed shouldn't be done here, tho if we ever reach to the point to allow DMs to replace vernacinterp as to take advantage of the data deps there, we could indeed implement this _à la Lean_ easily.

IMO we can keep the hack for abstract (and call the variable `command_loc_for_abstract` of something like that)

Note how the location will still be imprecise in the case of chained tactics), but I think we should remove all the other uses of the hack.

SkySkimmer replied:
It's not just for abstract, it's for any command whose implementor has other things to do than pass around locs

ejgallego replied:
Actually locs are pretty important IMO, so command implementors should care about them, but if you wanna go this way, OK then.

SkySkimmer replied:
>how would the developer discover that they need to override the global ref in order to pass the loc, if they want to override the loc?

They shouldn't override the loc, it's "current command loc" not "current whatever loc"
If they want a more precise loc the ?loc argument needs to be added to the api

ejgallego replied:
Indeed, my worry is that how people looking at the API of commands will discover that this is handled in a global variable. Anyways IMHO I'm happy to end the current discussion, I wrote many years ago a CEP to have a more common vernac handling, this actually has uses in UIs, but not yet. We can retake this discussion if we need on the CEP.

SkySkimmer commented on vernac/prettyp.ml:
```diff
@@ -904,6 +904,23 @@ let canonical_info env ref =
+let loc_info gr =
+  match Nametab.cci_src_loc gr with
+  | None -> mt()
+  | Some loc -> cut() ++ hov 0 (str "Declared in" ++ spc() ++ pr_loc_use_dp loc)
```
not sure if we want to always print this, it could cause a lot of churn in the output tests

SkySkimmer commented:
@coqbot run full ci

ejgallego reviewed (changes requested):
The hack in declare is a recipe for problems, I don't see why it is needed. My request is to remove it.

Also note the comments here and in #16261 , but these are not blockers, tho this change may have an impact on users, as noted in some issues.

I'm not sure it is a good idea to have two sources of location for objects (libobject and .glob files)

@SkySkimmer you can also maybe compare to #16261 to see if we updated the same places.

gares commented:
> The hack with `Loc.get_current_loc` is a recipe for problems, I don't see why it is needed. My request for changes is to remove the hack.

@SkySkimmer  made the point that passing the loc in all code paths is invasive and in some cases not doable (eg abstract/ltac). So completely removing `Loc.get_current_loc` is not an option.

Could you please explain precisely the problems this ref poses? I find that the wrapper `Util.try_finally` makes it very hard to misuse the new ref, so I don't know what worries you @ejgallego.

ejgallego commented:
> Could you please explain precisely the problems this ref poses? I find that the wrapper `Util.try_finally` makes it very hard to misuse the new ref, so I don't know what worries you @ejgallego.

I have explained the problems I see with the ref above, and additionally, it can be problematic for example if we ever do multithread. I think most people here is aware of the problem globals ref pose when having to reason about the code, for robustness, and maintenance. I think it is safe to safe that lack of types in our interfaces have created large pain for users and developers alike in the past due to bugs.

Moreover, as original implementor of the declare API I think passing it explicitly is the right choice.

Do we agree that the current solution would be called "a hack"?

Following my point in the answer to Gaëtan, why to use this hack for loc, and not for other information?

In the end it is not a big deal, so feel free to go ahead with the way it is, but it would take little time to follow the style done in #16261 , it doesn't take so much work to have cleaner code.

SkySkimmer commented:
>Do we agree that the current solution would be called "a hack"?

no

SkySkimmer commented:
loc is particularly annoying to have to pass explicitly because it's optional so easy to forget

ejgallego commented:
> loc is particularly annoying to have to pass explicitly because it's optional so easy to forget

That's true of other parameters; loc is actually pretty important once you want to do incremental stuff by the way.

That being said, it would be nice if:
- loc coming from Ast would not be optional anymore
- we move a few `?loc` to `~loc`
